### PR TITLE
Add Pipe Commands extension

### PR DESCRIPTION
**Pull request by pomdtr**
## Description

Pipe commands allows you to send your selection to a script command, and act on it. Example usage:
- send selected file by mail
- transform user selection to lowercase
- search selected word in google...

Some scripts are provided as examples in the assets folder.

## Type of change

- New extension

## Screencast

Screen Recording 2022-01-25 at 00 26 06

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

### CODE REVIEW

**Comment by fturcheti** [comment]
Hey, @pomdtr! I was testing some icons with more complex forms, trying to indicate something being transformed or piped, but I didn't like the results. Icons in Raycast are too small, so simpler forms tend to fit better. 

This is the one I liked the most: 

icon-pipe-to-command

What do you think?

**Comment by fturcheti** [comment]
From my understanding of the way this extensions is designed, it is not possible to create a command (a pipe command) that presents a list of actions to the user, right? 

The use case I'm thinking of is the one of my extension "Open With App". It works like this:

1. You select a file in Finder (or multiple files)
2. Launch Raycast
3. Select the command "Open With…"
4. Select an application from a list of applications to open the file with

**Comment by pomdtr** [comment]
I love the icon, thanks a lot !
I have decided to hardcode the openWithApp action :)

**Comment by fturcheti** [comment]
> I actually want to create another extension allowing to generate list from script commands, using a syntax inspired by alfred script filters.

Nice!

**Comment by pomdtr** [comment]
Not sure why the CI returns an error here, `npm run lint` does work on my side.

**Comment by fturcheti** [comment]
Hey, @pomdtr. When using pipe commands for texts transformations, I'd prefer the result to be copied to the clipboard instead of direct pasted. 

I know this can be changed by editing the commands source code, but I'd like to know your thoughts on which should be the suggested default for this transformations. I think that putting the transformed text in the clipboard is more flexible—you can `cmd+v` it to substitute the original text or use it whatever you want, because you have it "saved" on your clipboard.

**Comment by pomdtr** [comment]
> Hey, @pomdtr. When using pipe commands for texts transformations, I'd prefer the result to be copied to the clipboard instead of direct pasted. 
> 
> I know this can be changed by editing the commands source code, but I'd like to know your thoughts on which should be the suggested default for this transformations. I think that putting the transformed text in the clipboard is more flexible—you can `cmd+v` it to substitute the original text or use it whatever you want, because you have it "saved" on your clipboard.

I agree, I think that all text commands should allow to either replace the selection or go to the clipboard (with a pref to configure the default action).

We should never have to use pbcopy in the script code.

**Comment by pomdtr** [comment]
> @pomdtr I'll list here other suggestions/thoughts:
> 
> * It would be nice to have a "Edit Pipe Command" command that would list your commands with a default action to "Edit" them with your default text editor.
> * Besides acting on the selection, it could fallback to act on the clipboard.
> * Is it possible to edit or remove built-in pipe commands? This options are only shown for the user created ones.
> * If I start to rename a file in Finder and invoke the command with the filename selected, the extension detects the selection as "file" not as "text". This is not the expected behavior.

I agree, I think the extension will just copy the default scripts to the supportFolder on the first run (all pipe commands will be editable / deletable).

**Comment by pomdtr** [comment]
> If I start to rename a file in Finder and invoke the command with the filename selected, the extension detects the selection as "file" not as "text". This is not the expected behavior.

Unfortunately I did not find a way to correct this with raycast current API. The current behavior is the best I was able to achieve.

**Comment by fturcheti** [comment]
> Unfortunately I did not find a way to correct this with raycast current API. The current behavior is the best I was able to achieve.

I understand. But this got me thinking: maybe it'd be a good idea to have a "Pipe Clipboard to Command" command. I'm not sure if I'd use more "pipe from clipboard" or "pipe from selection", but "pipe from clipboard" would work for a situation like this file renaming—I could copy the name, invoke "Pipe Clipboard to Command", apply a pipe command to transform the text, and get the result in my clipboard to paste over the original name.

Would a "Pipe Clipboard to Command" work for files too?

**Comment by pomdtr** [comment]
The extension is currently defaulting to the clipboard if no item is selected. But maybe I should introduce a real `pipe clipboard` command ?

**Comment by fturcheti** [comment]
> The extension is currently defaulting to the clipboard if no item is selected. But maybe I should introduce a real pipe clipboard command ?

Maybe instead of using the clipboard as a fallback in the "Pipe Selection to Command", it's a better idea to have a "Pipe Clipboard to Command"—it's more explicit on what's going on. Also, there are situations like this:

1. You select some text
2. Invoke the "Pipe Selection to Command"
3. The selection detection fails (on an unsupported app, or for whatever reason)
4. The extension use the clipboard as a fallback
5. The extension pastes the transformed text from the clipboard over the original selection, although the user would like to transform the selection content, not the clipboard one

I know I have suggested the clipboard as a fallback, but thinking about it a bit more, it doesn't seem to be a good idea. Another command, to explicitly act on the clipboard seems to be a better approach.

What do you think? Does it make sense to you?

_
P.S. I just discovered that Raycast can't detect Neovim selections on the terminal, and that's my editor of choice. So I'd use the "Pipe Clipboard to Command" more than I initially thought.

**Comment by pomdtr** [comment]
Ok this should be ready for review !
@fturcheti I would love to get your feedback on the last version.

**Comment by fturcheti** [comment]
Hey, @pomdtr! I'll keep testing the extension, but here are some things that already popped up: 

- I've got an "Invalid Pipe Command" for a ".DS_Store" file (screenshot below). I think the fix for it is just a matter of filtering them out of this listing, right?
- "Pipe Clipboard to Command" is not working with files. Is this a limitation from the API?
- I miss an "Edit Pipe Command" that would list all your pipe commands (maybe in with list sections for each type—text, file and url). It would make it easy to manage them. 
- It'd be hand to have a shortcut assigned  to the "Pipe Output" action (maybe `⌘ + ]`)

Screen Shot 2022-01-31 at 23 30 24

**Comment by pomdtr** [comment]
Hey @fturcheti thanks for the feedback !

1. Yeah I think all dotfiles should be ignored
2. I have no way to find out that the current clipboard is a file using the pbpaste command :/
3. I agree. I think I will remove the invalid pipe command from the default menu, and display them here
4. I agree, I think I will use `cmd+shift+\` (close to a shell pipe `|`)

**Comment by pernielsentikaer** [comment]
What also could be an awesome idea, is that the first launch showed a `detail view` with information on how to use it somehow

**Comment by fturcheti** [comment]
@pomdtr Another thing that I noticed makes me less confident with the UX: it doesn't make it clear what's the thing your piping into a command. To tackle this with the current UI components from the API, I see two options:

1. Use a List section title
2. Or a Toast with the style ToastStyle.Animated

This component would display the text (or part of the text…) you have selected, or the URL, or the filename, depending on the thing your piping. If there's nothing selected, it could display a notice while still making the list available for managing pipe commands — an error message would appear only if the user tried to run an action on an empty selection.

What do you think?

**Comment by fturcheti** [comment]
@pomdtr And just a small UI tweak: you could use the "globe" icon for URL commands, to differentiate them from text commands.

**Comment by pomdtr** [comment]
Interesting idea. I removed the url type I felt like it did not bring enough values compared to the text type

**Comment by pomdtr** [comment]
I feel like it will be easier to show the user the selection once the dual panel will be available in the raycast API

**Comment by pomdtr** [comment]
> "Pipe Clipboard to Command" is not working with files. Is this a limitation from the API?

I found a way to achieve this using applescript :)

File: extensions/pipe-commands/src/pipe-to-command.tsx [hunk 1]
```diff
@@ -43,7 +43,7 @@ export function PipeCommands(props: { input: PipeInput }): JSX.Element {
       {props.input.type == "file" ? (
         <List.Item
           icon={Icon.Document}
-          title="Open With"
+          title="Open with..."
```
Comment (fturcheti) [inline extensions/pipe-commands/src/pipe-to-command.tsx:46]: I think it's better to name it "Open With…", as this is the way Raycast name this action in the built-in File Search extension.

File: extensions/pipe-commands/src/manage-pipe-commands.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    parseScriptCommands().then(setState);
+  };
+
+  useEffect(loadCommands, []);
+  const textCommands = state?.commands.filter((command) => command.metadatas.input.type === "text");
+  const fileCommands = state?.commands.filter((command) => command.metadatas.input.type === "file");
+
+  return (
+    <List isLoading={typeof state == "undefined"}>
+      <List.Section title="text">
```
Comment (fturcheti) [inline extensions/pipe-commands/src/manage-pipe-commands.tsx:21]: I think "Text Commands" would be a better name for the section

File: extensions/pipe-commands/src/manage-pipe-commands.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        {textCommands?.map((cmd) => (
+          <PipeCommand key={cmd.path} command={cmd} reload={loadCommands} />
+        ))}
+      </List.Section>
+      <List.Section title="file">
```
Comment (fturcheti) [inline extensions/pipe-commands/src/manage-pipe-commands.tsx:26]: The same way, "File Commands" would fit better here

File: extensions/pipe-commands/src/manage-pipe-commands.tsx [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  const loadCommands = () => {
+    parseScriptCommands().then(setState);
+  };
+
+  useEffect(loadCommands, []);
+  const textCommands = state?.commands.filter((command) => command.metadatas.input.type === "text");
+  const fileCommands = state?.commands.filter((command) => command.metadatas.input.type === "file");
+
+  return (
+    <List isLoading={typeof state == "undefined"}>
```
Comment (fturcheti) [inline extensions/pipe-commands/src/manage-pipe-commands.tsx:20]: In regards to the order of the sections, I'd use this order: "invalid", "file", "text". I'd display "invalid" at the top so the user could notice it and act to solve the issue with these commands.  And I'd put the "file" section before the "text" one cause I think these should be alphabetic ordered.

**Comment by pomdtr** [comment]
Thanks for the feedback @fturcheti 
I adressed all of these :)

**Comment by fturcheti** [comment]
Hey @pomdtr! What do you think about custom icons for pipe commands? Take a look:

Screen Shot 2022-02-16 at 09 41 08

Screen Shot 2022-02-16 at 09 41 21

**Comment by fturcheti** [comment]
> I added this feature in https://github.com/raycast/extensions/commit/db57c0b95776f8fbf6fe5146da137f0698fa9daf :)

Nice @pomdtr ! I saw your comment suggesting a PR but I was not able to do it properly at that time. Also, I'm a completely newbie with Typescript. I'm starting to "get it" looking though these Raycast extensions, so it's nice to see how you solved this.

In regards to the icons itself, the good thing about using the built-in ones is that they keep things clean, concise, and they work with both light and dark themes. When providing external icons, you can end up with contrast problems (the Github favicon is a problematic one, as it will disapear in the dark theme). Script-commands, for instance, have an "icon" and an "iconDark" property to tackle this issue.

**Comment by pomdtr** [comment]
> > I added this feature in https://github.com/raycast/extensions/commit/db57c0b95776f8fbf6fe5146da137f0698fa9daf :)
> 
> Nice @pomdtr ! I saw your comment suggesting a PR but I was not able to do it properly at that time. Also, I'm a completely newbie with Typescript. I'm starting to "get it" looking though these Raycast extensions, so it's nice to see how you solved this.
> 
> In regards to the icons itself, the good thing about using the built-in ones is that they keep things clean, concise, and they work with both light and dark themes. When providing external icons, you can end up with contrast problems (the Github favicon is a problematic one, as it will disapear in the dark theme). Script-commands, for instance, have an "icon" and an "iconDark" property to tackle this issue.

You are right, I will add support for the iconDark property.

**Comment by fturcheti** [comment]
And just another comment in regard to this: in my dirty hack I kept the different default icons for each type of command. So if a `text` command didn't have a custom icon, it would default to `Icon.Text`, and a `file` command would default to `Icon.Document`. This seems to be a good thing to keep from the implementation prior to custom icons. With a generic default we loose this basic differentiation between the types of commands. What do think?

**Comment by pomdtr** [comment]
> And just another comment in regard to this: in my dirty hack I kept the different default icons for each type of command. So if a text command didn't have a custom icon, it would default to Icon.Text, and a file command would default to Icon.Document. This seems to be a good thing to keep from the implementation prior to custom icons. With a generic default we loose this basic differentiation between the types of commands. What do think?

Great suggestion ! Fixed in b479fc2

**Comment by thomaspaulmann** [comment]
What's the latest on this? Looks like folks can't wait for it ;-) 

https://twitter.com/srikat/status/1496569084154384385

**Comment by pomdtr** [comment]
The extension is waiting for review. I stil do some minor improvements from time to time as I'm using it daily.

**Review by mathieudutour — COMMENTED** [review]
It looks amazing, really good job!

I just have a comment about supporting `@raycast.argument1` instead of `@raycast.input` so that some Script Commands can work out of the box with this extension.

Let me know if you want to do that, otherwise I'll merge it

**Comment by mathieudutour** [comment]
Another idea I had while playing with it: it'd be nice to have "pipelines", some sort of scripts than call some others in row

**Comment by pomdtr** [comment]
> Another idea I had while playing with it: it'd be nice to have "pipelines", some sort of scripts than call some others in row

I think this usecase should be handled directly by the shell by calling multiple other pipe commands from the pipe command.

**Comment by pomdtr** [comment]
> It looks amazing, really good job!

> I just have a comment about supporting @raycast.argument1 instead of @raycast.input so that some Script Commands can work out of the box with this extension.

> Let me know if you want to do that, otherwise I'll merge it

I added support for the `raycast.argument1` field, and dropped `raycast.input`. For convenience, the input is passed both as stdin and as the script first argument.

**Review by mathieudutour — APPROVED** [review]
Neat, let's go 🚀

**Pull request merged**